sdk/go/pulumi/types.go

```diff
@@ -282,6 +283,10 @@ func (o *OutputState) getState() *OutputState {
 }
 
 func newOutputState(join *workGroup, elementType reflect.Type, deps ...Resource) *OutputState {
+	if deps == nil && len(deps) != 0 {
```

Comment:
Implementing @AaronFriel suggestion here - in our analysis so far, the torn slice must enter this module through newOutputState, because mergeDependencies is observing a torn slice in first argument, and it's the only way this can happen presumably.. So while not a fix, this will catch the issue with a more helpful stack trace helping us understand where the torn slice is originating from hopefully.